fe/estimator_abfe.py

```diff
@@ -0,0 +1,301 @@
+
+
+FreeEnergyModel = namedtuple(
+    "FreeEnergyModel",
+    [
+     "unbound_potentials",
+     "endpoint_correct",
+     "client",
+     "box",
+     "x0",
+     "v0",
+     "integrator",
+     "barostat",
+     "lambda_schedule",
+     "equil_steps",
+     "prod_steps",
+     "beta",
+     "prefix",
+    ]
+)
+
+gradient = List[Any] # TODO: make this more descriptive of dG_grad structure
+
+def _deltaG(model, sys_params) -> Tuple[Tuple[float, List], np.array]:
+
+    assert len(sys_params) == len(model.unbound_potentials)
+
+    bound_potentials = []
+    for params, unbound_pot in zip(sys_params, model.unbound_potentials):
+        bp = unbound_pot.bind(np.asarray(params))
+        bound_potentials.append(bp)
+
+    all_args = []
+    for lamb_idx, lamb in enumerate(model.lambda_schedule):
+
+        if model.endpoint_correct and lamb_idx == len(model.lambda_schedule) - 1:
+            x_interval = 200
+        else:
+            x_interval = 10000
+
+        all_args.append((
+            lamb,
+            model.box,
+            model.x0,
+            model.v0,
+            bound_potentials,
+            model.integrator,
+            model.barostat,
+            model.equil_steps,
+            model.prod_steps,
+            x_interval
+        ))
+
+    if model.endpoint_correct:
+        all_args.append((
+            1.0,
+            model.box,
+            model.x0,
+            model.v0,
+            bound_potentials[:-1],
+            model.integrator,
+            model.barostat,
+            model.equil_steps,
+            model.prod_steps,
+            200
+        ))
+
+    if model.client is None:
+        results = []
+        for args in all_args:
+            results.append(simulate(*args))
+    else:
+        futures = []
+        for args in all_args:
+            futures.append(model.client.submit(simulate, *args))
+
+        results = []
+        for future in futures:
+            results.append(future.result())
+
+    mean_du_dls = []
+    all_grads = []
+
+    if model.endpoint_correct:
+        ti_results = results[:-1]
+    else:
+        ti_results = results
+
+    for lambda_idx, (lambda_window, result) in enumerate(zip(model.lambda_schedule, ti_results)):
+        # (ytz): figure out what to do with stddev(du_dl) later
+        print(f"{model.prefix} index {lambda_idx} lambda {lambda_window:.5f} <du/dl> {np.mean(result.du_dls):.5f} med(du/dl) {np.median(result.du_dls):.5f}  o(du/dl) {np.std(result.du_dls):.5f}")
+        mean_du_dls.append(np.mean(result.du_dls))
+        all_grads.append(result.du_dps)
+
+    tibar_dG = 0
+    for lambda_idx in range(len(model.lambda_schedule) - 1):
+        fwd_work = ti_results[lambda_idx].du_dls
+        rev_work = -ti_results[lambda_idx+1].du_dls
+        delta_lamb = model.lambda_schedule[lambda_idx+1] - model.lambda_schedule[lambda_idx]
+        tibar, err = pymbar.BAR(model.beta*fwd_work, model.beta*rev_work)
+        tibar_dG += (tibar/model.beta)*delta_lamb
```

Comment:
Documentation request: add comment at start of this block describing intent and any notable approximations, possibly with a reference. Alternately, move this free energy estimator into a separate function.

Some approximations I think should be noted:
* Estimate free energy difference between lam[0] and lam[-1] by summing up BAR estimates for subsequent pairs of states lam[i], lam[i+1]. Requires as input "delta U"s, i.e. `w_F=u(x, lam[i+1]) - u(x, lam[i])`.
* First-order Taylor approximation for `w_F=u(x, lam[i+1]) - u(x, lam[i])`, approximated as `w_F_prime=delta_lamb * du_dl(x, lam[i])`
* Interchange order of operations, replacing `BAR(w_F_prime, w_R_prime)` with `delta_lamb * BAR(w_F_prime/delta_lamb, w_R_prime/delta_lamb)`. (Probably doesn't make a huge difference, but should be noted that `b * BAR(w_F/b, w_R/b) != BAR(w_F, w_R)` in general.)

Reply:
Willfix

Reply:
PS I haven't decided if I should include this yet, adding this here right now mainly for diagnostic purposes.

Reply:
Resolved in https://github.com/proteneer/timemachine/pull/452/commits/a0dbff7c9b7327d12c2354397d9e5172cab61bb5 and https://github.com/proteneer/timemachine/pull/452/commits/212550143194be4583576cfce17a1af26474b5b4